python/pylib/pose.py

```diff
@@ -20,912 +16,1317 @@ def quatInit():
 
-def mul_Quat_Quat_Array ( q1, q2 ):
-    result = np.empty((np.shape(q1)[0],4))
+    result = np.empty((max(n1, n2), 4))
     result[:,0] = q1[:,0]*q2[:,0] - q1[:,1]*q2[:,1] - q1[:,2]*q2[:,2] - q1[:,3]*q2[:,3]
     result[:,1] = q1[:,0]*q2[:,1] + q1[:,1]*q2[:,0] - q1[:,2]*q2[:,3] + q1[:,3]*q2[:,2]
     result[:,2] = q1[:,0]*q2[:,2] + q1[:,1]*q2[:,3] + q1[:,2]*q2[:,0] - q1[:,3]*q2[:,1]
     result[:,3] = q1[:,0]*q2[:,3] - q1[:,1]*q2[:,2] + q1[:,2]*q2[:,1] + q1[:,3]*q2[:,0]
 
+    if array == 0:
+        result = np.squeeze(result)
+    return result
+
+
+# * Product of two Quaternions.  Order of q1 and q2 matters (same as applying two successive DCMs)!!!
+# * Combines two quaternion rotations into one rotation.
+# * result = quatConj(q1) * q2
+# * Reference: http://www.mathworks.com/help/aeroblks/quaternionmultiplication.html
+def mul_ConjQuat_Quat(q1, q2):
+    if len(np.shape(q1)) == 1:
+        q1 = np.expand_dims(q1, axis=0)
+        array = 0
+    else:
+        array = 1
+    if len(np.shape(q2)) == 1:
+        q2 = np.expand_dims(q2, axis=0)
+    else:
+        array = 1
+    n1 = np.shape(q1)[0]
+    n2 = np.shape(q2)[0]
+    assert n1 == n2 or n1 == 1 or n2 == 1, "Number of quaternions in arrays do not match"
+
+    result = np.empty((max(n1, n2), 4))
+    result[:,0] = q1[:,0]*q2[:,0] + q1[:,1]*q2[:,1] + q1[:,2]*q2[:,2] + q1[:,3]*q2[:,3]
+    result[:,1] = q1[:,0]*q2[:,1] - q1[:,1]*q2[:,0] + q1[:,2]*q2[:,3] - q1[:,3]*q2[:,2]
+    result[:,2] = q1[:,0]*q2[:,2] - q1[:,1]*q2[:,3] - q1[:,2]*q2[:,0] + q1[:,3]*q2[:,1]
+    result[:,3] = q1[:,0]*q2[:,3] + q1[:,1]*q2[:,2] - q1[:,2]*q2[:,1] - q1[:,3]*q2[:,0]
+
+    if array == 0:
+        result = np.squeeze(result)
+    return result
+
+
+# * Product of two Quaternions.  Order of q1 and q2 matters (same as applying two successive DCMs)!!!
+# * Combines two quaternion rotations into one rotation.
+# * result = q1 * quatConj(q2)
+# * Reference: http://www.mathworks.com/help/aeroblks/quaternionmultiplication.html
+def mul_Quat_ConjQuat(q1, q2):
+    if len(np.shape(q1)) == 1:
+        q1 = np.expand_dims(q1, axis=0)
+        array = 0
+    else:
+        array = 1
+    if len(np.shape(q2)) == 1:
+        q2 = np.expand_dims(q2, axis=0)
+    else:
+        array = 1
+    n1 = np.shape(q1)[0]
+    n2 = np.shape(q2)[0]
+    assert n1 == n2 or n1 == 1 or n2 == 1, "Number of quaternions in arrays do not match"
+
+    result = np.empty((max(n1, n2), 4))
+    result[:,0] =  q1[:,0]*q2[:,0] + q1[:,1]*q2[:,1] + q1[:,2]*q2[:,2] + q1[:,3]*q2[:,3]
+    result[:,1] = -q1[:,0]*q2[:,1] + q1[:,1]*q2[:,0] + q1[:,2]*q2[:,3] - q1[:,3]*q2[:,2]
+    result[:,2] = -q1[:,0]*q2[:,2] - q1[:,1]*q2[:,3] + q1[:,2]*q2[:,0] + q1[:,3]*q2[:,1]
+    result[:,3] = -q1[:,0]*q2[:,3] + q1[:,1]*q2[:,2] - q1[:,2]*q2[:,1] + q1[:,3]*q2[:,0]
+
+    if array == 0:
+        result = np.squeeze(result)
     return result
 
 
 #  * Division of two Quaternions.  Order matters!!!
```

Comment:
Initially I thought it's the same as multiplying one quaternion by inverse of the other. But it's not. I'm not sure how this function is supposed to be used.

Reply:
If the math looks suspicious that is a red flag to me